library/pkcs12.c

```diff
@@ -219,17 +321,14 @@ int mbedtls_pkcs12_derivation( unsigned char *data, size_t datalen,
 
     unsigned char diversifier[128];
     unsigned char salt_block[128], pwd_block[128], hash_block[128] = {0};
-    unsigned char hash_output[MBEDTLS_MD_MAX_SIZE];
+    unsigned char hash_output[64]; /* Maximal hash size for SHA512 */
```

Comment:
Why we can't use `MBEDTLS_MD_MAX_SIZE` here?
I see we are now using the hard-coded maximal size (`64`) and I remember that there was problem with this buffer in tests. Is this the same case?

Reply:
I think that's a good candidate for a small refactoring task later on.

Reply:
Of course the trivial way of solving just this instance of the problem would be
```c
#if defined(MBEDTLS_MD_C)
unsigned char hash_output[MBEDTLS_MD_MAX_SIZE];
#else
unsigned char hash_output[PSA_HASH_MAX_SIZE];
```

Reply:
The code will get cluttered with such definitions, IMHO it's better to have a common layer above that would resolve it :)

Reply:
Shouldn't this be more like:
```
#if defined(MBEDTLS_MD_C) && (                  \
    !defined(MBEDTLS_PSA_CRYPTO_C) ||           \
    MBEDTLS_MD_MAX_SIZE > PSA_HASH_MAX_SIZE )
#define MBEDTLS_HASH_MAX_SIZE MBEDTLS_MD_MAX_SIZE
#elif defined(MBEDTLS_PSA_CRYPTO_C) && (            \
    !defined(MBEDTLS_MD_C) ||                     \
    PSA_HASH_MAX_SIZE > MBEDTLS_MD_MAX_SIZE )
#define MBEDTLS_HASH_MAX_SIZE PSA_HASH_MAX_SIZE
#else
#define MBEDTLS_HASH_MAX_SIZE MBEDTLS_MD_MAX_SIZE
#endif
```

Reply:
> IMHO it's better to have a common layer above that would resolve it

I know I must sound like a broken record, but it still looks to me like that common layer should be `md.h`.

Reply:
@mprse Indeed my suggestion was not correct, I should be using `>=` rather than `>` and an `#elif` in the middle to avoid a double definition.

However, not having an `#else` was intentional: if neither `MD_C` nor `PSA_CRYPTO_C` is defined, then there's no way to compute a hash, so any use of `HASH_MAX_SIZE` is probably an error and I'd rather get an error about it.

That said, it's probably better to leave those details to a later task, not just to avoid extending the scope of this one, but also because it will be easier to resolve when we can look at all the use cases at once. So, I've created #6252